traffic_portal/test/integration/config.model.ts

```diff
@@ -0,0 +1,165 @@
+ * under the License.
+ */
+import { hasProperty } from "./CommonUtils/utils";
+
+/** Possible levels of TO Alerts */
+export type AlertLevel = "success" | "info" | "warning" | "error";
+
+/**
+ * Checks whether an object is a valid Alert level.
+ *
+ * @param s The object to check.
+ * @returns Whether or not `s` is an AlertLevel.
+ */
+export function isAlertLevel(s: unknown): s is AlertLevel {
+    if (typeof(s) !== "string") {
+        return false;
+    }
+    switch (s) {
+        case "success":
+        case "info":
+        case "warning":
+        case "error":
+            return true;
+    }
+    return false;
+}
+
+/** TO API Alerts */
+export interface Alert {
+    level: AlertLevel;
+    text: string;
+}
+
+/**
+ * Checks whether an object is an Alert like the ones TO normally returns.
+ *
+ * @param a The object to check.
+ * @returns Whether or not `a` is an Alert.
+ */
+export function isAlert(a: unknown): a is Alert {
+    if (typeof(a) !== "object" || a === null) {
+        return false;
+    }
+    if (!hasProperty(a, "level") || !hasProperty(a, "text", "string")) {
+        return false;
+    }
+    return isAlertLevel(a.level);
+}
+
+/**
+ * Logs an alert to the appropriate console stream based on its `level`.
+ *
+ * @param a The Alert to log.
+ * @param prefix Optional prefix for the log message
+ */
+export function logAlert(a: Alert, prefix?: string): void {
+    let logfunc;
+    let pre = (prefix ?? "").trimStart();
+    switch (a.level) {
+        case "success":
+            logfunc = console.log;
+            pre = `SUCCESS: ${pre}`;
+            break;
+        case "info":
+            logfunc = console.info
+            pre = `INFO: ${pre}`;
+            break;
+        case "warning":
+            logfunc = console.warn
+            pre = `WARN: ${pre}`;
+            break;
+        case "error":
+            logfunc = console.error
+            pre = `ERROR: ${pre}`;
+            break;
+    }
+    logfunc(pre, a.text);
+}
+
+/** TestingConfig is the type of a testing configuration. */
+export interface TestingConfig {
+	/** This is login information for a user with admin-level permissions. */
+	readonly login: {
+		readonly password: string;
+		readonly username: string;
+	};
+	/** The URL at which the Traffic Ops API can be accessed. */
+	readonly apiUrl: string;
+	/** The URL at which Traffic Portal is served - root path. */
+	readonly baseUrl: string;
+	/** Logging alert levels that are enabled. */
+	readonly alertLevels?: Array<AlertLevel>;
+}
+
+/**
+ * Checks if a passed object is a valid testing configuration.
+ *
+ * @param c The object to check.
+ * @returns `true`, always. When the check fails, it throws an error that
+ * explains why.
+ */
+export function isTestingConfig(c: unknown): c is TestingConfig {
+	if (typeof(c) !== "object") {
+		throw new Error(`testing configuration must be an object, not a '${typeof(c)}'`);
+	}
+	if (c === null) {
```

Comment:
nit: null check before `typeof` is marginally faster in case of nulls

Reply:
Yeah, but `typeof` before null check is marginally faster in case of anything that isn't `null` or an object, which is more things than just `null`. I think it'd be more common for the configuration to have a typo or be incomplete so that the input value is actually `undefined` like e.g.:
```json
{
    "arams": {
        "config": "options here"
    }
}
```
... than it would be for it to be like:
```json
{
    "params": null
}
```

Reply:
Fair point